scanpy/metrics/_morans_i.py

```diff
@@ -0,0 +1,263 @@
+    obsm: Optional[str] = None,
+    obsp: Optional[str] = None,
+    use_raw: bool = False,
+) -> Union[np.ndarray, float]:
+    r"""
+    Calculate Moran’s I Global Autocorrelation Statistic.
+
+    Moran’s I is a global autocorrelation statistic for some measure on a graph. It is commonly used in
+    spatial data analysis to assess autocorrelation on a 2D grid. It is closely related to Geary's C,
+    but not identical. More info can be found `here`<https://en.wikipedia.org/wiki/Moran%27s_I> .
+
+    .. math::
+
+        I=\frac{n}{S_{0}} \frac{\sum_{i=1}^{n} \sum_{j=1}^{n} w_{i, j} z_{i} z_{j}}{\sum_{i=1}^{n} z_{i}^{2}}
+
+    Params
+    ------
+    adata
+    vals
+        Values to calculate Moran's I for. If this is two dimensional, should
+        be of shape `(n_features, n_cells)`. Otherwise should be of shape
+        `(n_cells,)`. This matrix can be selected from elements of the anndata
+        object by using key word arguments: `layer`, `obsm`, `obsp`, or
+        `use_raw`.
+    use_graph
+        Key to use for graph in anndata object. If not provided, default
+        neighbors connectivities will be used instead.
+    layer
+        Key for `adata.layers` to choose `vals`.
+    obsm
+        Key for `adata.obsm` to choose `vals`.
+    obsp
+        Key for `adata.obsp` to choose `vals`.
+    use_raw
+        Whether to use `adata.raw.X` for `vals`.
+
+
+    This function can also be called on the graph and values directly. In this case
+    the signature looks like:
+
+    Params
+    ------
+    g
+        The graph
+    vals
+        The values
+
+
+    See the examples for more info.
+
+    Returns
+    -------
+    If vals is two dimensional, returns a 1 dimensional ndarray array. Returns
+    a scalar if `vals` is 1d.
+
+
+    Examples
+    --------
+
+    Calculate Morans I for each components of a dimensionality reduction:
+
+    .. code:: python
+
+        import scanpy as sc, numpy as np
+
+        pbmc = sc.datasets.pbmc68k_processed()
+        pc_c = sc.metrics.morans_i(pbmc, obsm="X_pca")
+
+
+    It's equivalent to call the function directly on the underlying arrays:
+
+    .. code:: python
+
+        alt = sc.metrics.morans_i(pbmc.obsp["connectivities"], pbmc.obsm["X_pca"].T)
+        np.testing.assert_array_equal(pc_c, alt)
+    """
+    if use_graph is None:
+        # Fix for anndata<0.7
+        if hasattr(adata, "obsp") and "connectivities" in adata.obsp:
+            g = adata.obsp["connectivities"]
+        elif "neighbors" in adata.uns:
+            g = adata.uns["neighbors"]["connectivities"]
+        else:
+            raise ValueError("Must run neighbors first.")
+    else:
+        raise NotImplementedError()
+    if vals is None:
+        vals = _get_obs_rep(adata, use_raw=use_raw, layer=layer, obsm=obsm, obsp=obsp).T
+    return morans_i(g, vals)
+
+
+@njit(
+    ft(ft[:], it[:], it[:], ft[:], ft[:], ft, ft),
+    cache=True,
+    parallel=False,
+    fastmath=True,
```

Comment:
Does this change anything?

I'm not sure if it's good to have on here. I'd be more amenable if this were an approximate method (like nndecent), but feel like statistic computation should probably be precise.

Reply:
actually,
```
# w/ fastmath
10.6 s ± 146 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
# w/o fastmath
12 s ± 1.17 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

again, not much, if you want I can remove.

Reply:
I'm more concerned about if the results changed. But I'm also not super compelled by the benchmark because the difference between the means is ~ the sum of the std deviations.

I don't know a ton about how this can change results, so I'd like to err on the side of caution and remove it.

Reply:
done